ext/drx/scatter_gather_aarch64.c

```diff
@@ -481,47 +493,60 @@ expand_scatter_gather(void *drcontext, instrlist_t *bb, instr_t *sg_instr,
                 orig_app_pc));
 
-        const reg_id_t reg_to_extract =
-            reg_is_z(sg_info->base_reg) ? sg_info->base_reg : sg_info->index_reg;
-
-        /* lastb    scratch_gpr0, scratch_pred, reg_to_extract.element_size */
-        EMIT(lastb_sve_scalar, opnd_create_reg(scratch_gpr0),
-             opnd_create_reg(scratch_pred),
-             opnd_create_reg_element_vector(reg_to_extract, sg_info->element_size));
-
-        const reg_id_t scalar_index_or_base = scratch_gpr0;
+        for (uint reg_index = 0; reg_index < sg_info->reg_count; reg_index++) {
+            if (reg_index == 0) {
+                const reg_id_t reg_to_extract =
+                    reg_is_z(sg_info->base_reg) ? sg_info->base_reg : sg_info->index_reg;
+
+                /* lastb    scalar_index_or_base, scratch_pred,
+                 * reg_to_extract.element_size */
+                EMIT(lastb_sve_scalar, opnd_create_reg(scalar_index_or_base),
+                     opnd_create_reg(scratch_pred),
+                     opnd_create_reg_element_vector(reg_to_extract,
+                                                    sg_info->element_size));
+            } else {
+                // Increment index
+                EMIT(add, opnd_create_reg(scalar_index_or_base),
+                     opnd_create_reg(scalar_index_or_base), OPND_CREATE_INT(1));
+            }
 
-        if (sg_info->is_load) {
-            const reg_id_t scalar_dst = scratch_gpr0;
-
-            /* ldr[bh]  scalar_dst, [mem] */
-            instrlist_preinsert(
-                bb, sg_instr,
-                INSTR_XL8(create_scalar_load_or_store(drcontext, sg_info,
-                                                      scalar_index_or_base, scalar_dst),
-                          orig_app_pc));
-
-            /* cpy      gather_dst_reg.element_size, scratch_pred/m, scalar_dst */
-            EMIT(cpy_sve_pred,
-                 opnd_create_reg_element_vector(sg_info->gather_dst_reg,
-                                                sg_info->element_size),
-                 opnd_create_predicate_reg(scratch_pred, true),
-                 opnd_create_reg(reg_resize_to_opsz(scalar_dst, sg_info->element_size)));
-        } else {
-            const reg_id_t scalar_src = scratch_gpr1;
-
-            /* lastb    scalar_src, scratch_pred, scatter_src_reg.element_size */
-            EMIT(lastb_sve_scalar, opnd_create_reg(scalar_src),
-                 opnd_create_reg(scratch_pred),
-                 opnd_create_reg_element_vector(sg_info->scatter_src_reg,
-                                                sg_info->element_size));
-
-            /* str[bh]  scalar_src, [mem] */
-            instrlist_preinsert(
-                bb, sg_instr,
-                INSTR_XL8(create_scalar_load_or_store(drcontext, sg_info,
-                                                      scalar_index_or_base, scalar_src),
-                          orig_app_pc));
+            if (sg_info->is_load) {
```

Comment:
If it's not too much refactoring, could we move the iteration logic from this point onwards into its own separate routine?

Reply:
See my response above.